### PR TITLE
Optimization rule for limit push down

### PR DESCRIPTION
**Pull request by jievince**
NG-239

### CODE REVIEW

File: src/optimizer/rule/LimitPushDownRule.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    if (!gnFound) {
+        return std::make_pair(false, matched);
+    }
+
+    const auto gn = static_cast<const GetNeighbors *>(gnExpr->node());
+    if (limitRows >= gn->limit()) {
+        return std::make_pair(false, matched);
+    }
+    matched.push_back(projExpr);
+    matched.push_back(gnExpr);
```
Comment (yixinglu) [inline src/optimizer/rule/LimitPushDownRule.cpp:135]: Returned OptGroupExprs only contains the root GroupExpr rather than all touched GroupExprs

Comment (jievince) [reply src/optimizer/rule/LimitPushDownRule.cpp:135]: Why? If just return projExpr, function LimitPushDownRule::transform will execute the following code again:

>     bool gnFound = false;
    const OptGroupExpr *gnExpr = nullptr;
    for (auto dep : projExpr->dependencies()) {
        for (auto expr : dep->groupExprs()) {
            if (expr->node()->kind() == PlanNode::Kind::kGetNeighbors) {
                gnFound = true;
                gnExpr = expr;
                break;
            }
        }
    }

Comment (yixinglu) [reply src/optimizer/rule/LimitPushDownRule.cpp:135]: make sense. I'm thinking about whether need to collect these OptGroupExprs by patten match.

**Review by yixinglu — APPROVED** [review]
I think this PR could be merged at present, I will update it when OptRule interface changed.

File: src/optimizer/rule/LimitPushDownRule.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        newGn->setStatProps(std::move(statPropsPtr));
+    }
+
+    if (gn->exprs()) {
+        auto exprs = *gn->exprs();
+        auto exprsPtr = std::make_unique<decltype(exprs)>(std::move(exprs));
+        newGn->setExprs(std::move(exprsPtr));
+    }
+    return newGn;
+}
```
Comment (CPWstatic) [inline src/optimizer/rule/LimitPushDownRule.cpp:197]: Emm, these `clone`s should be provided by the node itself.

**Review by CPWstatic — DISMISSED** [review]

**Comment by jievince** [comment]
clone function is fixed in yee's pr https://github.com/vesoft-inc/nebula-graph/pull/311/files

File: src/optimizer/rule/LimitPushDownRule.cpp (hunk 2 above)
Comment (jievince) [reply src/optimizer/rule/LimitPushDownRule.cpp:197]: this is fixed in yee's pr https://github.com/vesoft-inc/nebula-graph/pull/311/files

**Review by yixinglu — DISMISSED** [review]
Good job!

File: src/optimizer/rule/LimitPushDownRule.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    auto projExpr = matched.dependencies.front().node;
+    auto gnExpr = matched.dependencies.front().dependencies.front().node;
+
+    const auto limit = static_cast<const Limit *>(limitExpr->node());
+    const auto proj = static_cast<const Project *>(projExpr->node());
+    const auto gn = static_cast<const GetNeighbors *>(gnExpr->node());
+
+    int64_t limitRows = limit->offset() + limit->count();
+    if (limitRows >= gn->limit()) {
+        return TransformResult{false, false, {}};
```
Comment (CPWstatic) [inline src/optimizer/rule/LimitPushDownRule.cpp:57]: It's difficult to understand, suggest add some static method for such situations, such as
```
TransformResult::emptyResult()
Or 
TransformResult::noTrans()
```

File: src/optimizer/OptRule.h [hunk 4]
```diff
@@ -50,6 +50,10 @@ class OptRule {
         bool eraseCurr{false};
         bool eraseAll{false};
         std::vector<OptGroupExpr *> newGroupExprs;
+        static const TransformResult &noTransform() {
+            static TransformResult kNoTrans{false, false, {}};
+            return kNoTrans;
+        }
```
Comment (CPWstatic) [inline src/optimizer/OptRule.h:56]:
```
static method -> method -> static variable -> variable
```
Emm, the static method should be place at the beginning of the class/struct.

Comment (jievince) [reply src/optimizer/OptRule.h:56]: Thanks

**Review by yixinglu — APPROVED** [review]

**Review by yixinglu — DISMISSED** [review]

**Review by CPWstatic — DISMISSED** [review]

**Review by laura-ding — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**